doc/src/errorrecovery.md

```diff
@@ -1,8 +1,396 @@
+
+```
+>>> 2+
+thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Lexeme { start: 2, len: 4294967295, tok_id: 4 }', libcore/result.rs:1009:5
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+>>> 
+```
+
+In this case, the first repair sequence was `Insert INT`. The fundamental
+problem is that while error recovery can adjust the user’s input to insert a
+lexeme of type `INT`, neither it nor the parser have any idea what content might
+have made sense for that lexeme. Thus the expression above caused the expression
+`$lexer.lexeme_str(&$1.unwrap())` to panic, since `$1` was `Err(<lexeme>)`.
+
+It is thus up to the user to decide what to do in the face of the inevitable
+semantic issues that error recovery highlights. Fortunately, this is generally
+simpler than it sounds with only a slight rethink in the way that we tend to
+write a grammar's actions.
+
+
+## A rule of thumb: make `%type` return a `Result` type
+
+Although you can use whatever type you use for `%type`, using a `Result` type
+allows a (deliberately) simple interaction with the effects of error recovery.
+The basic idea is simple: in actions, we ignore lexemes whose content we don't
+care about (e.g. brackets); for lexemes whose content we care about, we either
+introduce a default value, or percolate an `Err` upwards. Default values make
+sense in certain situations. For example, if you're writing a compiler, and want
+to run a static type checker even after syntax errors, it might make sense to
+assume that `Insert 0` is a good substitute for `Insert INT`. However, in the
+case of the calculator, default values are likely to lead to confusing results.
+We thus change the grammar so that inserted integers prevent evaluation from
+occurring:
+
+```yacc
+%start Expr
+%type Result<u64, ()>
+%%
+Expr: Term 'PLUS' Expr { Ok($1? + $3?) }
+    | Term { $1 }
+    ;
+
+Term: Factor 'MUL' Term { Ok($1? * $3?) }
+    | Factor { $1 }
+    ;
+
+Factor: 'LBRACK' Expr 'RBRACK' { $2 }
+      | 'INT' {
+            let l = $1.map_err(|_| ())?;
+            Ok(parse_int($lexer.lexeme_str(&l)))
+         }
+      ;
+%%
+fn parse_int(s: &str) -> u64 {
+    match s.parse::<u64>() {
+        Ok(val) => val as u64,
+        Err(_) => panic!("{} cannot be represented as a u64", s)
+    }
+}
+```
+
+The basic idea here is that every action returns an instance of `Result<u64,
+()>`: if we receive `Ok(u64)` we successfully evaluated the expression, but if
+we received `Err(())` we were not able to evaluate the expression. If we
+encounter an integer lexeme which is the result of error recovery, then the
+`INT` lexeme in the second `Factor` action will be `Err(<lexeme>)`. By writing
+`$1.map_err(|_| ())?` we’re saying “if the integer lexeme was created by error
+recovery, percolate `Err(())` upwards”. We then have to tweak a couple of other
+actions to percolate errors upwards, but this is a trivial change.
+
+We then need to make a small tweak to our `main.rs` changing:
+
```

Comment:
You could present this as a diff, e.g.:
```diff,ignore
match res {
-    Some(r) => println!("Result: {}", r),
+    Some(Ok(r)) => println!("Result: {}", r),
    _ => eprintln!("Unable to evaluate expression.")
}
```

Reply:
As discussed in person, mdbook doesn't seem to colour diffs in the way we would like.